site/content/blog/2020-06-10-opentracing.md

```diff
@@ -0,0 +1,619 @@
+---
+title: "OpenTracing for Go Projects"
+slug: open-tracing
+series: "AST"
+date: 2020-06-10T00:00:00-04:00
+author: Eli Yukelzon
+github: reflog
+community: eli.yukelzon
+toc: true
+---
+
+## What is distributed tracing?
+
+Large scale, cloud applications are usually built using interconnected services that can be rather hard to troubleshoot. When a service is scaled, simple logging doesn't cut it anymore and a more in-depth view into system's flow is required.
+That's where [Distributed tracing](https://opentracing.io/docs/overview/what-is-tracing/) comes into play; it allows developers and SREs to get a detailed view of a request as it travels through the system of services. With distributed tracing you can:
+
+1. Trace the execution path of a single request as it goes through a complicated path inside the distributed system
+2. Pinpoint bottlenecks and measure latency of specific parts of the execution path
+3. Record and analyze system behavior
+
+[OpenTracing](https://opentracing.io) is an open standard describing how distributed tracing works.
+
+There are a few key terms used in tracing:
+
+1. **Trace**: A recording of the execution path of a request
+2. **Span**: A named, timed operation representing a contiguous segment inside the trace
+3. **Root Span**: The first span in a trace - a common ancestor to all spans in a trace
+4. **Context**: Information identifying the request, required to connect spans in a distributed trace
+
+A trace recording usually looks something like this:
+
+![trace image](/blog/2020-06-10-opentracing/trace.png) 
+
+We've previously explored an OpenTracing implementation in the [first post of this series]({{< relref "./2019-10-25-instrumenting-go-code-via-ast.md" >}}). Next, we want to add distributed tracing capabilities to [mattermost-server](https://github.com/mattermost/mattermost-server). For this, we've picked [OpenTracing Go](https://github.com/opentracing/opentracing-go). 
+
+In this article we'll discuss all the nitty-gritty details of implementing a tracing system in your Go application without littering your code with repetitive, boilerplate tracing code.
+
+## The goal
+
+So what are we actually working on? We want to make sure that every API request that's being handled by our server will get recorded into a trace, together with context information. This gives us the ability to dive deep into the execution and allow easy problem analysis.
+
+The resulting system trace will look like this (using [Jaeger](https://www.jaegertracing.io/) web-ui visualization):
+![jaeger view](/blog/2020-06-10-opentracing/result.png) 
+
+## Naive approach
```

Comment:
```suggestion
## Naive approach
```
I wonder if we could rename this to "Simple tracing implementation"? Reason for this is because we don't refer to a "Sophisticated approach" as a counterpart. 

I'm 0/5 on the replacement word/term, but 4/5 on changing it.

Reply:
@justinegeffen what do you think about "Straightforward tracing implementation"? Because I don't know if "simple" works here...

Reply:
That works well!